crates/pink/src/runtime/extension.rs

```diff
@@ -53,11 +58,120 @@ pub fn get_side_effects() -> ExecSideEffects {
 pub struct PinkExtension;
 
 impl ChainExtension<super::PinkRuntime> for PinkExtension {
-    fn call<E: Ext>(func_id: u32, _env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    fn call<E: Ext>(func_id: u32, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
     where
         <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
     {
-        error!(target: "pink", "Called an unregistered `func_id`: {:}", func_id);
-        return Err(DispatchError::Other("Unimplemented func_id"));
+        let call = Call { env };
+        // func_id refer to https://github.com/patractlabs/PIPs/blob/main/PIPs/pip-100.md
+        match func_id {
+            0xff000001 => call.http_request(),
+            _ => {
+                error!(target: "pink", "Called an unregistered `func_id`: {:}", func_id);
+                return Err(DispatchError::Other("Unimplemented func_id"));
+            }
+        }
+    }
+}
+
+struct Call<'a, 'b, E: Ext> {
+    env: Environment<'a, 'b, E, InitState>,
+}
+
+impl<'a, 'b, E: Ext> Call<'a, 'b, E>
+where
+    <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+    fn http_request(self) -> Result<RetVal, DispatchError> {
+        use pink_extension::chain_extension::{HttpRequest, HttpResponse};
+        if !matches!(get_call_mode(), Some(CallMode::Query)) {
+            return Err(DispatchError::Other(
+                "http_request can only be called in query mode",
+            ));
+        }
+
+        let mut env = self.env.buf_in_buf_out();
+        let request: HttpRequest = env.read_as_unbounded(env.in_len())?;
+        let uri = http_req::uri::Uri::try_from(request.url.as_str())
+            .or(Err(DispatchError::Other("Invalid URL")))?;
+
+        let mut req = http_req::request::Request::new(&uri);
+        for (key, value) in &request.headers {
+            req.header(key, value);
+        }
+
+        match request.method.as_str() {
+            "GET" => {
+                req.method(http_req::request::Method::GET);
+            }
+            "POST" => {
+                req.method(http_req::request::Method::POST)
+                    .body(request.body.as_slice());
+            }
+            _ => {
+                return Err(DispatchError::Other("Unsupported method".into()));
+            }
+        };
+
+        // Hardcoded limitations for now
+        const MAX_WAIT_TIME: u64 = 10; // seconds
```

Comment:
Ideally we should allow the cluster owner to config the max timeout?

Reply:
Should consider it in the future.